rotkehlchen/chain/ethereum/modules/makerdao/sai/decoder.py

```diff
@@ -0,0 +1,524 @@
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.user_messages import MessagesAggregator
+
+POOLED_ETHER_ADDRESS = string_to_evm_address('0xf53AD2c6851052A81B42133467480961B2321C09')
+MAKERDAO_SAITUB_CONTRACT = string_to_evm_address('0x448a5065aeBB8E423F0896E6c5D525C040f59af3')
+MAKERDAO_SAITAP_CONTRACT = string_to_evm_address('0xBda109309f9FafA6Dd6A9CB9f1Df4085B27Ee8eF')
+MAKERDAO_SAI_PROXY_CONTRACT = string_to_evm_address('0x526af336D614adE5cc252A407062B8861aF998F5')
+MAKERDAO_SAITUB_NEW_CDP_TOPIC = b'\x89\xb8\x89;\x80m\xb5\x08\x97\xc8\xe26,qW\x1c\xfa\xeb\x97a\xee@r\x7fh?\x17\x93\xcd\xa9\xdf\x16'  # noqa: E501
+MAKERDAO_SAITUB_CLOSE_CDP_TOPIC = b'\xb8M!\x06\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'  # noqa: E501
+MAKERDAO_SAITUB_BORROW_SAI_TOPIC = b'D\x0f\x19\xba\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'  # noqa: E501
+MAKERDAO_SAITUB_REPAY_SAI_TOPIC = b's\xb3\x81\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'  # noqa: E501
+MAKERDAO_SAITUB_DEPOSIT_WETH_TOPIC = b'\x04\x98x\xf3\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'  # noqa: E501
+MAKERDAO_SAITUB_LIQUIDATE_TOPIC = b'@\xcc\x88T\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'  # noqa: E501
+MAKERDAO_SAITUB_WITHDRAW_TOPIC = b'\xa5\xcd\x18N\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'  # noqa: E501
+
+
+class MakerdaosaiDecoder(DecoderInterface):
+    """
+    Information relating to event topics and contracts were gotten from:
+    https://github.com/makerdao/sai/blob/d28fe5074af7c661fb3aa7f117ad419fb6414b91/DEVELOPING.md
+    """
+    def __init__(
+            self,
+            ethereum_inquirer: 'EthereumInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(ethereum_inquirer, base_tools, msg_aggregator)
+        self.ethereum_inquirer = ethereum_inquirer
+        self.base_tools = base_tools
+        self.weth = A_WETH.resolve_to_evm_token()
+        self.eth = A_ETH.resolve_to_crypto_asset()
+        self.sai = A_SAI.resolve_to_evm_token()
+        self.peth = Asset(f'eip155:1/erc20:{POOLED_ETHER_ADDRESS}')
+
+    def _decode_makerdao_sai_tub(
+            self,
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],
+            action_items: Optional[list[ActionItem]],
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
+        if tx_log.topics[0] == MAKERDAO_SAITUB_NEW_CDP_TOPIC:
```

Comment:
Maybe a dictionary that maps topics to methods could be used here?

Reply:
Won't that be a case of an unneeded abstraction? To be clear, you meant something like this right ⬇️  
```python3
TOPICS_TO_METHODS = {
     MAKERDAO_SAITUB_NEW_CDP_TOPIC: '_decode_makerdao_new_cdp_event',
     MAKERDAO_SAITUB_CLOSE_CDP_TOPIC: '_decode_makerdao_close_cdp',
     MAKERDAO_SAITUB_BORROW_SAI_TOPIC: '_decode_makerdao_borrow_sai_event',
     MAKERDAO_SAITUB_REPAY_SAI_TOPIC: '_decode_makerdao_repay_sai_event',
}
try:
    decoder_method = getattr(self, TOPICS_TO_METHODS[tx_log.topics[0]])
    return decoder_method(
        tx_log=tx_log,
        transaction=transaction,
        decoded_events=decoded_events,
        all_logs=all_logs,
        action_items=action_items,
    )
except KeyError:
    return None, []

Reply:
Yeah, this.

Maybe it will ... let's see what Lefteris says

Reply:
Yes do it. It's not a bad idea. It will help reduce those huge if/elifs. I guess if it's more than 3 conditions it may start to make sense. Otherwise it would be an overkills.